jenkins/deploy-clio-server.sh

```diff
@@ -0,0 +1,209 @@
+set -ex
+
+# program
+PROG_NAME=$(basename $0)
+CLIO_DIR=$(cd $(dirname $(dirname $0)) && pwd)
+
+# This script is run on jenkins server in order to perform all the necessary
+# steps deploy a clio-server instance to a host.
+
+# For instance:
+#   build and push a docker image for clio-server
+#   gather configurations associated with app to deploy
+#   render any ctmpls
+#   copy configs to node
+#   stop any currently running container
+#   re-start container using new configs
+
+# the script expects the following ENV vars set to non-null values
+#  ENV: environment to deploy to
+#  CLIO_HOST: hostname to deploy clio to
+
+# the script also supports the optional ENV vars
+#  DOCKER_TAG: docker tag to apply to the built and deployed instance
+#              if not specified, the default SNAP version is used
+#  RUN_TESTS: if "false", unit tests will not be run as part of building
+#             the docker image
+
+# verify required vars are set to actual values
+if [ -z "$ENV" ]
+then
+   echo "Error: ENV not set!!"
+   exit 2
+fi
+
+if [ -z "$CLIO_HOST" ]
+then
+   echo "Error: CLIO_HOST not set!!"
+   exit 2
+fi
+
+CONFIG_DIR=clio-server/config
+
+if [ ! -d "$CONFIG_DIR" ]
+then
+    echo "Missing configuration directory ($CONFIG_DIR) - Exiting "
+    exit 1
+fi
+
+# set VAULT token location
+if [ -f /etc/vault-token-dsde ]
+then
+   TOKEN_FILE="/etc/vault-token-dsde"
+elif [ -f ${HOME}/.vault-token ]
+then
+   TOKEN_FILE="${HOME}/.vault-token"
+else
+   echo "Missing VAULT TOKEN file - Exitting"
+   exit 1
+fi
+   
+# initialize vars used in multiple config templates
+# location on host where configs should get copied
+APP_DIR=${APP_DIR:-"/app"}
+# location in the clio container where configs will be mounted
+CLIO_CONF_DIR=/etc
+# location in the clio container where logs will be mounted
+CLIO_LOG_DIR=/logs
+# name of the application config used by the clio instance
+CLIO_APP_CONF=clio.conf
+# name of the logback config used by the clio instance
+CLIO_LOGBACK_CONF=clio-logback.xml
+# name of the env file used to set up clio with docker-compose
+CLIO_ENV_FILE=clio.env
+# port to expose for clio on the host VM
+HOST_CLIO_PORT=80
+# port to expose for clio in the clio container
+CONTAINER_CLIO_PORT=${HOST_CLIO_PORT}
+
+# SSH options
+SSH_USER=jenkins
+SSHOPTS="-o UserKnownHostsFile=/dev/null -o CheckHostIP=no -o StrictHostKeyChecking=no"
+SSHCMD="ssh $SSHOPTS"
+SCPCMD="scp $SSHOPTS"
+
+# sbt command to disable tests in the docker build
+if [[ "$RUN_TESTS" = "false" ]]; then
+    SET_TESTS="set test in assembly in LocalProject(\"clio-server\") := {}"
+fi
+# sbt command to override the default docker tag
+if [[ ! -z "$DOCKER_TAG" ]]; then
+    SET_VERSION="set version := \"$DOCKER_TAG\""
+fi
+
+# build the clio-server docker image
+docker run --rm \
+    -v /var/run/docker.sock:/var/run/docker.sock \
+    -v ${CLIO_DIR}:/clio \
+    -w="/clio" \
+    broadinstitute/scala:scala-2.11.8 sbt "$SET_TESTS" "$SET_VERSION" "clio-server/docker"
```

Comment:
Using this docker image (instead of the Jenkins SBT plugin) allows us to pass environment variables as arguments, but it also slows down the deploy process because it doesn't have any libraries for SBT 0.13.15 or Scala 2.12 pre-cached. I opened DSDEGP-1425 to track publishing and using a Docker image with the Scala and SBT versions used by Clio.